quic/s2n-quic-transport/src/space/rx_packet_numbers/pending_ack_ranges.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{interval_set, space::rx_packet_numbers::ack_ranges::AckRanges};
+use core::time::Duration;
+use s2n_quic_core::{
+    frame::ack::EcnCounts,
+    packet::number::{PacketNumber, PacketNumberRange},
+};
+
+/// Stores ACK ranges pending processing
+#[derive(Clone, Debug, Default)]
+pub struct PendingAckRanges {
+    ranges: AckRanges,
+    ecn_counts: EcnCounts,
+    ack_delay: Duration,
+}
+
+impl PendingAckRanges {
+    #[inline]
+    pub fn new(ranges: AckRanges, ecn_counts: EcnCounts, ack_delay: Duration) -> Self {
+        PendingAckRanges {
+            ranges,
+            ecn_counts,
+            ack_delay,
+        }
+    }
+
+    /// Extend with a packet number range; dropping smaller values if needed
+    #[inline]
+    pub fn extend(
+        &mut self,
+        acked_packets: PacketNumberRange,
+        ecn_counts: Option<EcnCounts>,
+        ack_delay: Duration,
+    ) -> bool {
+        if let Some(ecn_counts) = ecn_counts {
+            self.ecn_counts = self.ecn_counts.max(ecn_counts);
+        }
+        // TODO: at the moment only a single payload(single delivery) worth of ACKs is
+        // batched for processing. This means that its acceptable to take the max
+        // ack_delay value.
+        //
+        // Once multiple payloads are stored/batched, multiple ack_delays might need to
+        // be stored.
+        self.ack_delay = self.ack_delay.max(ack_delay);
+
+        // TODO: add metrics if ack ranges are being dropped
+        self.ranges.insert_packet_number_range(acked_packets)
+    }
+
+    /// Returns an iterator over all of the values contained in the ranges `IntervalSet`.
+    #[inline]
+    pub fn iter(&self) -> interval_set::Iter<PacketNumber> {
+        self.ranges.iter()
+    }
```

Comment:
I actually think we'll want the [intervals iterator](https://github.com/aws/s2n-quic/blob/c3b3c3294cd0a90c2964faea0e5b41ded62e957f/quic/s2n-quic-transport/src/interval_set/mod.rs#L713) instead since the `ack::Set` allows for it.